mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,929 @@
-    /// Convert a slice into an index array
-    pub fn from_slice(
-        slice: &[u8],
-        used_entries: u32,
-    ) -> Result<Vec<MetaDataIndexEntry>, MetaDataError> {
-        let mut index: Vec<MetaDataIndexEntry> = Vec::new();
-        let mut reader = Cursor::new(slice);
-        for _ in 0 .. used_entries {
-            index.push(
-                deserialize_from(&mut reader).context(DeserializeError {})?,
-            );
+impl MetaDataIndex {
+    pub const METADATA_INDEX_SIZE: u32 = size_of::<MetaDataIndex>() as u32;
+    pub const INDEX_SIGNATURE: [u8; 8] =
+        [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61];
+
+    /// Convert a slice into a MetaDataIndex and validate
+    pub fn from_slice(slice: &[u8]) -> Result<MetaDataIndex, MetaDataError> {
+        let mut index: MetaDataIndex =
+            deserialize_from(&mut Cursor::new(slice))
+                .context(DeserializeError {})?;
+
+        if index.signature != MetaDataIndex::INDEX_SIGNATURE {
+            return Err(MetaDataError::IndexSignature {});
+        }
+
+        if index.index_size != MetaDataIndex::METADATA_INDEX_SIZE {
+            return Err(MetaDataError::IndexSize {});
+        }
+
+        let checksum = index.self_checksum;
+
+        if checksum != index.checksum().context(SerializeError {})? {
+            return Err(MetaDataError::IndexChecksum {});
         }
+
         Ok(index)
     }
 
-    /// Calculate checksum of an index array
-    pub fn checksum(index: &[MetaDataIndexEntry]) -> u32 {
-        let mut digest = crc32::Digest::new(crc32::IEEE);
-        for entry in index {
-            digest.write(&serialize(entry).unwrap());
+    /// Checksum the index with the checksum field itself set to 0
+    pub fn checksum(&mut self) -> Result<u32, Error> {
+        self.self_checksum = 0;
+        self.self_checksum = crc32::checksum_ieee(&serialize(self)?);
+        Ok(self.self_checksum)
+    }
+
+    pub fn new(
+        parent: Guid,
+        guid: Guid,
+        index_lba: u64,
+        total_entries: u64,
+    ) -> MetaDataIndex {
+        MetaDataIndex {
+            signature: MetaDataIndex::INDEX_SIGNATURE,
+            index_size: MetaDataIndex::METADATA_INDEX_SIZE,
+            self_checksum: 0,
+            parent,
+            guid,
+            generation: 0,
+            timestamp: 0,
+            self_lba: index_lba,
+            start_lba: index_lba + 1,
+            current_entry: 0,
+            used_entries: 0,
+            total_entries,
         }
-        digest.sum32()
     }
 }
 
-#[derive(Debug, PartialEq, Default, Clone)]
-pub struct NexusMetaData {
-    pub header: MetaDataHeader,
-    pub index: Vec<MetaDataIndexEntry>,
-}
+pub struct NexusMetaData;
 
 impl NexusMetaData {
-    /// Construct a MetaDataHeader from raw data
-    fn read_header(buf: &DmaBuf) -> Result<MetaDataHeader, MetaDataError> {
-        MetaDataHeader::from_slice(buf.as_slice())
+    fn timestamp(time: &SystemTime) -> u128 {
```

Comment:
This function does not interact with `NexusMetaData` at all, it should probably be removed from this trait into a bare namespace.